lib/quicly.c

```diff
@@ -4483,15 +4774,216 @@ static int handle_ping_frame(quicly_conn_t *conn, struct st_quicly_handle_payloa
+ */
+static int verify_new_cid(uint64_t sequence, const quicly_cid_t *cid,
+                          const uint8_t stateless_reset_token[QUICLY_STATELESS_RESET_TOKEN_LEN],
+                          const quicly_new_connection_id_frame_t *frame, int *retcode)
+{
+    /* If an endpoint receives a NEW_CONNECTION_ID frame that repeats a previously issued connection ID with
+     * a different Stateless Reset Token or a different sequence number, or if a sequence number is used for
+     * different connection IDs, the endpoint MAY treat that receipt as a connection error of type PROTOCOL_VIOLATION.
+     * (19.15)
+     */
+    if (cid_is_equal(cid, frame->cid) == 0) {
+        if (sequence == frame->sequence &&
+            memcmp(stateless_reset_token, frame->stateless_reset_token, QUICLY_STATELESS_RESET_TOKEN_LEN) == 0) {
+            /* likely a duplicate due to retransmission */
+            *retcode = 0;
+            return 1;
+        } else {
+            /* received a frame that carries conflicting information */
+            *retcode = QUICLY_TRANSPORT_ERROR_PROTOCOL_VIOLATION;
+            return 1;
+        }
+    }
+    if (sequence == frame->sequence && cid_is_equal(cid, frame->cid) != 0) {
+        *retcode = QUICLY_TRANSPORT_ERROR_PROTOCOL_VIOLATION;
+        return 1;
+    }
+
+    return 0;
+}
+
 static int handle_new_connection_id_frame(quicly_conn_t *conn, struct st_quicly_handle_payload_state_t *state)
 {
+    int ret;
     quicly_new_connection_id_frame_t frame;
-    return quicly_decode_new_connection_id_frame(&state->src, state->end, &frame);
+    int is_expected = 0; /* are we expected to receive this sequence no? i.e. we haven't retired this */
+
+    /* TODO: return error when using zero-length CID */
+
+    if ((ret = quicly_decode_new_connection_id_frame(&state->src, state->end, &frame)) != 0)
+        return ret;
+
+    QUICLY_PROBE(NEW_CONNECTION_ID_RECEIVE, conn, probe_now(), frame.sequence, frame.retire_prior_to,
+                 QUICLY_PROBE_HEXDUMP(frame.cid.base, frame.cid.len),
+                 QUICLY_PROBE_HEXDUMP(frame.stateless_reset_token, QUICLY_STATELESS_RESET_TOKEN_LEN));
+
+    /* conditions for the sequence numbers to be treated as an error or ignored
+     *
+     * If the received sequence number exceeds largest_sequence_expected (after considering the CIDs to be retired via
+     * retire_prior_to), it is an ACTIVE_CONNECTION_LIMIT error.
+     * If the number is not found in the list of expected sequence numbers (conn->super.peer.spare_cids), it is considered to be
+     * already retired. */
+
+    uint64_t num_to_retire = frame.retire_prior_to > conn->super.peer.largest_retire_prior_to
+                                 ? frame.retire_prior_to - conn->super.peer.largest_retire_prior_to
+                                 : 0;
+    if (frame.sequence > conn->super.peer.largest_sequence_expected + num_to_retire) {
```

Comment:
I'm not sure if this condition is true.

Consider the case where `largest_retire_prior_to` is zero and `frame.retire_prior_to` is 4. RHS of this comparison expression yields 4, and therefore we'd call it a protocol error when `frame.sequence` is any value above 4. Is that correct? I do not think so. As our local limit is 4, it is my understanding that sequence numbers up to 7 have to be considered valid.

Regarding how we should fix this issue, I think it might make sense to remove this check altogether. Instead, we could change the structure of this function to make changes in following steps:
1. Retire CIDs as instructed. At this point, the spare slots would be assigned a new sequence number, with the `is_active` being set to false. This entire logic can be coded in `schedule_retire_connection_id`.
2. Store the newly provided CID. At this point, we'd be looking for a slot that has the matching sequence number. If we could not find a slot with a matching sequence number, we know that the peer has sent something invalid.

Reply:
PS. To paraphrase, there are two operations that modify "the list of CIDs provided by the peer":
* when a potentially new CID is given
* when a CID is retired (regardless of the reason; e.g., us switching to a different address, peer using NCID.RPT)
Having two functions dedicated for each operation might make the code cleaner.

Reply:
> Consider the case where largest_retire_prior_to is zero and frame.retire_prior_to is 4. RHS of this comparison expression yields 4, and therefore we'd call it a protocol error when frame.sequence is any value above 4. Is that correct? I do not think so. As our local limit is 4, it is my understanding that sequence numbers up to 7 have to be considered valid.

Hmm I'm not sure if we are on the same page on how this works.

Let's say the host is announcing `active_connection_id_limit = 4` and got 4 CIDs (including the initial one at the connection establishment). At this moment `largest_sequence_expected == 3`. And as in your example, let's say largest_retire_prior_to is zero and frame.retire_prior_to is 4. Then this expression
```
conn->super.peer.largest_sequence_expected + num_to_retire
```
gives 3+4=7. Thus `frame.sequence == 7` will be accepted as a valid number.

Reply:
As we discussed offline, I'm more than happy to refactor it so that we can achieve the same goal with smaller LOC or much simpler code. So let me understand your idea more thoroughly.

As you pointed out probably I can get rid of the comparison `if (frame.sequence > conn->super.peer.largest_sequence_expected + num_to_retire)`. The only reason I put it upfront was that I was reluctant to modify the state (e.g. `spare_cids`) if the frame is going to be determined invalid, but since we'll close the connection in case of invalid frame and dispose all of per-connection states, maybe we don't have to worry about it.

> Retire CIDs as instructed. At this point, the spare slots would be assigned a new sequence number, with the is_active being set to false. This entire logic can be coded in schedule_retire_connection_id.

To be precise I'm skeptical about it. Currently `schedule_retire_connection_id` is used in many other places such as retiring the current CID for communication and rescheduling RCID after packet loss, in which cases there exists no corresponding `spare_cid_t` entry.

> To paraphrase, there are two operations that modify "the list of CIDs provided by the peer": ... (snip) ... Having two functions dedicated for each operation might make the code cleaner.

I think the "new CID is given" part is `store_spare_cid` function. I agree that it would be better to have a counterpart for deactivating the spare (i.e. setting `is_active` to false and set the next expected sequence number). Is that what you are suggesting?

Reply:
> The only reason I put it upfront was that I was reluctant to modify the state (e.g. `spare_cids`) if the frame is going to be determined invalid, but since we'll close the connection in case of invalid frame and dispose all of per-connection states, maybe we don't have to worry about it.

:+1:

>> Retire CIDs as instructed. At this point, the spare slots would be assigned a new sequence number, with the is_active being set to false. This entire logic can be coded in schedule_retire_connection_id.
> 
> To be precise I'm skeptical about it. Currently `schedule_retire_connection_id` is used in many other places such as retiring the current CID for communication and rescheduling RCID after packet loss, in which cases there exists no corresponding `spare_cid_t` entry.

:+1:

>> To paraphrase, there are two operations that modify "the list of CIDs provided by the peer": ... (snip) ... Having two functions dedicated for each operation might make the code cleaner.
> 
> I think the "new CID is given" part is store_spare_cid function. I agree that it would be better to have a counterpart for deactivating the spare (i.e. setting is_active to false and set the next expected sequence number). Is that what you are suggesting?

Kind of. That matches what I'd expect the `unregister_cid` method to do in the example below. But let's step back for a moment.

My concern is that this is a complex stuff. That is due to the spec being complex, the people involved in the standardization (including me) are to blame.

I am very happy to see this PR having all the necessary code that implements the complex spec, thank you very much for your efforts. Now might be a good time to start refactoring our design to make sure that we have all things correctly laid out, and so that it could be maintained without the fear of breaking something.

Considering that, I think it would be a good idea to split frame processing and state processing. Also, if we could possibly create an object (or objects) that retains the states, it would be easier to write unit tests just regarding the state processing.

Regarding how an ideal code might look like, we might think of something like below.

```c++
/**
 * Retains active CIDs.
 */
struct ActiveCIDs {
    struct {
        uint64_t sequence;
        quicly_cid_t cid;
        uint8_t srt[...];
    } active_cids[4];
    /**
     * Initializez the object.
     */
    ActiveCIDs();
    /**
     * Registers a new CID.
     * Returns 0 if successful, otherwise a transport error.
     */
    int register_cid(uint64_t sequence, ptls_iovec_t cid, const void *srt);
    /**
     * Unregisters a CID, given the index of the CID within `active_cids`.
     * The caller sends an RCID frame after calling this function.
     */
    void unregister_cid(size_t cid_index);
}
```

This is just an example. This example assumes that we handle the current (in-flight) CID and the spares in the same way (we can, but we do not do that now).

Also, this object does not keep RPT by itself. That's because when the object is going to process RPT, it needs to signal retirement of multiple CIDs. However, that's going to require a callback, or having a way to pass a list of sequence numbers. We can do that, or we could retain the pending active CIDs _and pending _RCIDs_ in a single object. But as I was uncertain if that is a good design, I am providing here a simple example.

Reply:
PS. To be honest, if I were you, I would be worried about possibly breaking this PR if we are to do a refactor like this. And because I would be worried of that, I would create a new PR against this PR for that purpose, _if_ I would try to see if we can make the code simpler by refactoring the code.

Reply:
Thanks for providing ideas. Actually, to my embarrassment, it hasn't occurred to me that issued CIDs and received CIDs can be (and perhaps should be) handled in a common data type. Looks like a good idea. Let me dig into this direction and get back to you. Meanwhile I've pushed trivial fixes and cleanups.

Reply:
> This example assumes that we handle the current (in-flight) CID and the spares in the same way (we can, but we do not do that now).

Now I realized that I misunderstood this comment. So this is about unifying the current CID and spare ones, not unifying structures for issued CIDs and received CIDs. . Within the received CIDs (current CID and spare ones), I'll explore a way to unify them, as it'll definitely simplify the logic.

Regarding unifying issued CIDs and received CIDs, though these two are quite similar but since they are used in totally disjoint contexts, for the moment I'd keep them separate.

Reply:
@kazuho Just submitted hfujita/quicly#1 based on your idea here. Please take a look and let me know what you think. If this direction looks okay, I'm thinking about migrating `issued_cid` state processing logic & `retire_cid` state processing logic into `cid.c` as well.

Reply:
Thank you for working on the refactor. I must say that I like it! It has become easier to read, and we have tests!

> If this direction looks okay, I'm thinking about migrating `issued_cid` state processing logic & `retire_cid` state processing logic into `cid.c` as well.

I agree that having them isolated too would help. Just curious, do you expect to have interaction between the three components (i.e. between `receive_cid`, `issued_cid`, and `retire_cid`), or is it just a coincidence that we'd have the three in one file?

Reply:
Thank you for a quick turnaround! At this moment I'm not sure if we're to see interaction between them. I just thought it'd be nice to separate code from quickly.c as much as reasonably possible.

Reply:
I see, thank you for the response.

To clarify, my weak preference goes to not having interaction between them. And if we want to be firm about that, I think we can use different files (pair of .h and .c files) for each of the three.

I do not think we need to be firm at this point, though. We can always split files at a later moment. I merely wanted to know if your idea of moving more logic into one file was based on the assumption that we'd be having interaction between the components stored in that one file.